content/blog/debug-cran-additional.md

```diff
@@ -0,0 +1,374 @@
+
+```
+docker run -v $PWD:/src:ro -it --rm --security-opt seccomp=unconfined wch1/r-debug
+```
+
+then, in that container, I prepared the R library in that container using the R binary that was built with the undefined behaviour checkers enabled:
+
+```
+RDcsan -e 'install.packages(c("deSolve", "knitr", "ring", "microbenchmark", "rmarkdown", "testthat", "devtools", "roxygen2"))'
+cp -r /src /dde
+```
+
+Running
+
+```
+RDcsan CMD INSTALL --preclean --install-tests /dde
+RDcsan -e 'tools::testInstalledPackage("dde")'
+cat dde-tests/testthat.Rout
+```
+
+showed the same output as CRAN reported - confirming I could replicate the error but still obscuring which test triggered it.  In the end I ran the tests with the most verbose reporter:
+
+```
+RDcsan -e 'devtools::test("dde", reporter = testthat::LocationReporter)'
+```
+
+which produces
+
+```
+Start test: failure to fetch history
+dopri.c:745:21: runtime error: -12.7138 is outside the range of representable values of type 'unsigned long'
+    #0 0x7f472e7aa95a in dopri_find_time /dde/src/dopri.c:745:21
+    #1 0x7f472e7ab097 in ylag_1 /dde/src/dopri.c:769:24
+    #2 0x7f472cded83a in ylag_1 /dde/inst/include/dde/dde.c:15:10
+...
+    #250 0x7f474672e010 in getvar /tmp/r-source/src/main/eval.c:5128:14
+
+SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior dopri.c:745:21 in
+  test-dde.R#269:1 [success]
+End test: failure to fetch history
+```
+
+so we know where the error is triggered from!  (Note that the UB checker only seems to report the error the *first time* it occurs in a session - it's quite possible this is tuneable though.)
+
+With a little surgery, a standalone script that reproduces the error is:
+
+```
+res <- dde:::shlib(system.file("examples/seir.c", package = "dde"), "dde_")
+y0 <- c(1e7 - 1, 0, 1, 0)
+times <- seq(0, 30, length.out = 301)
+dde::dopri(y0, times, "seir", numeric(),
+           atol = 1e-7, rtol = 1e-7, n_history = 2L,
+           dllname = "dde_seir", return_history = FALSE)
+```
+
+and we can trigger the error in the instrumented copy of R by running `RDcsan -f /src/bug.R` in ~12s which is a bit less tedious than running the full suite. However, most of the debugging can be done in the plain copy.  I compiled the package with optimisation turned off and debugging symbols enabled (`CFLAGS = -g -O0`), ran R with `R -d gdb` and created a breakpoint with `break dopri.c:745` after loading the package.
+
+The relevant bit of C looks like:
+
+```
+  size_t idx0 = 0;
+  if (n > 0) {
+    const double
+      t0 = ((double*) ring_buffer_tail(obj->history))[idx_t],
+      t1 = ((double*) ring_buffer_tail_offset(obj->history, n - 1))[idx_t];
+    idx0 = min_size((t - t0) / (t1 - t0) / (n - 1), n - 1);
+  }
+```
+
+When triggered (the second assignment to `idx0`), we have (approximately) `t0 = 11`, `t1 = 12` and `t = 0.02` which falls outside of the range of times, so the expression `(t - t0) / (t1 - t0) / (n - 1)` is negative and that's the undefined behaviour.
```

Comment:
I can accept this but it's hard to follow when you're not sure what the code is meant to be doing, let alone what t0 and t1 are.

Reply:
thanks - @weshinsley was confused here too so I have expanded a bit